neps/nep-0539.md

```diff
@@ -0,0 +1,307 @@
+        # linear interpolation based on congestion level
+        outgoing_gas_limit[receiver]
+          = mix(30 Pgas, 0 Pgas, incoming_congestion(receiver))
+    ```
+2. (new) Drain receipts in the outgoing buffer from the previous round-
+    - Subtract `receipt.gas()` from `outgoing_gas_limit[receipt.receiver]` for
+      each receipt drained.
+    - Keep receipts in the buffer if the gas limit would be negative.
+    - Add the removed receipts to the outgoing receipts of the new chunk.
+3. Convert all transactions to receipts included in the chunk.
+    - Local receipts, which are receipts where the sender account id is equal to
+      the receiver id, are set aside as local receipts for step 4.
+    - Non-local receipts up to `outgoing_gas_limit[receipt.receiver]` for the
+      respective shard go to the outgoing receipts list of the chunk.
+    - (new) Non-local receipts above `outgoing_gas_limit[receipt.receiver]` for
+      the respective shard go to the outgoing receipts buffer.
+4. Execute receipts in the order of `local`, `delayed`, `incoming`.
+    - Don't stop before all receipts are executed or more than 1000 Tgas have
+      been burnt. Burnt gas includes the burnt gas from step 3.
+    - Outgoing receipts up to what is left in
+      `outgoing_gas_limit[receipt.receiver]` per shard (after step 3) go to the
+      outgoing receipts list of the chunk.
+    - (new) Outgoing receipts above `outgoing_gas_limit[receipt.receiver]`
+      go to the outgoing receipts buffer.
+5. Remaining local or incoming receipts are added to the end of the `delayed`
+   receipts queue.
+6. (new) Compute own congestion information for the next block.:
+    ```python
+    MAX_CONGESTION_INCOMING_GAS = 100 PGas
+    gas_backlog = sum([receipt.gas() for receipt in delayed_receipts_queue])
+    incoming_congestion = gas_backlog / MAX_CONGESTION_INCOMING_GAS
+    incoming_congestion = min(1.0, incoming_congestion)
+
+    MAX_CONGESTION_MEMORY_CONSUMPTION = 500 MB
+    memory_consumption = 0
+    memory_consumption += sum([receipt.size() for receipt in delayed_receipts_queue])
+    memory_consumption += sum([receipt.size() for receipt in postponed_receipts_queue])
+    memory_consumption += sum([receipt.size() for receipt in outgoing_receipts_buffer])
+  
+    general_congestion = memory_consumption / MAX_CONGESTION_MEMORY_CONSUMPTION
+    general_congestion = min(1.0, general_congestion)
+    ```
+
+*TODO: fine-tune `MAX_CONGESTION_INCOMING_GAS` and `MAX_CONGESTION_MEMORY_CONSUMPTION`*
+
+In the formula above, the receipt gas and the receipt size are defined as:
+
+```python
+def gas(receipt):
+  return receipt.attached_gas + receipt.exec_gas
+
+def size(receipt):
+  return len(borsh(receipt))
+```
+
+### Change to Trie
+
+We store the outgoing buffered receipts in the trie, similar to delayed receipts
+but in their own separate column. Therefore we add a trie column
+`BUFFERED_RECEIPT_OR_INDICES: u8 = 13;`. Then we read and write analogue to the
```

Comment:
Minor detail: I know we use this pattern for `DELAYED_RECEIPT_OR_INDICES`, but it seems to be that way for historical reasons ([see commit message here](https://github.com/near/nearcore/commit/7365af44942f984a23ae7c1956d9b96802e52c9b)).

For this new queue it would be clearer to have separate `BUFFERED_RECEIPT` and `BUFFERED_RECEIPT_INDICES` columns.